_posts/2016-07-11-value-of-automated-e2e-testing.md

```diff
@@ -0,0 +1,58 @@
+---
+layout: post
+author: riikkanen
+title: >
+  The Value of Automated e2e-testing in a Software Project
+excerpt: >
+  A peek into automated e2e-testing done in a software project. The first try with Wicket Page Tester. The second and more sophisticated version with JBehave and Selenium.
+tags:
+- Java
+- testing
+- e2e testing
+- JBehave
+- Wicket
+- Selenium
+- programming
+---
+
+The software project concerned started in the beginning of Autumn in 2010. I had the honor to work in the project as a full stack developer from its beginning to June 2014 when my maternity leave began. The project was about to renew a 30 years old, technically outdated registry and everything attached to it, such as the user inteface and all the integrations. The project was a success, after almost three years of hard work, the application was taken into serious use without any major hardships. The project is still alive and kicking, even though, nowadays mostly maintenance, but also some feature development. 
+
+##The Application
+
+The application is written in Java. The actual user interface of the registry is developed as a web application using Apache Wicket, which is an Java web application framework. The data is stored in Oracle database and queried with direct SQL-queries, without any intermediate layer such as Hibernate. There are also numerous integrations bubbling under, powered by Mule. 
+
+The views in the user interface are designed in such a way that the user can see the main item and all the sub-items related at the same time. The main item could have several sub-items listed in a table-like manner. The main item has certain managing tools and actions and, of course, all the sub-items have a bunch of tools and actions too. Most of the tools are implemented as icons with actions attached. The state and visibility of the tools depend about the status of the main item, but also from the context. Is this the main item which created the sub-item, the main item which edited them or a main item having nothing to do with the others? In addition to, some notification icons need to be shown in certain situations. However, after all we had a number of different icons which should be visible if and only if the sunspots were in the right position. 
```

Comment:
The views are described very accurately. If this serves an important point would an image be necessary?